shell/platform/android/external_view_embedder/external_view_embedder.cc

```diff
@@ -299,8 +299,11 @@ bool AndroidExternalViewEmbedder::SupportsDynamicThreadMerging() {
 }
 
 // |ExternalViewEmbedder|
-void AndroidExternalViewEmbedder::Teardown() {
-  surface_pool_->DestroyLayers(jni_facade_);
+void AndroidExternalViewEmbedder::Teardown(
+    fml::RefPtr<fml::RasterThreadMerger> raster_thread_merger) {
+  if (raster_thread_merger->IsOnPlatformThread()) {
+    surface_pool_->DestroyLayers(jni_facade_);
+  }
```

Comment:
So if it's called on the wrong thread it's a silent no-op? Is that going to leak things?

Reply:
this is an extra protection to match similar calls to this method. 
However, `shell/common/rasterizer.cc` is the actual fix that prevents the leak and the issue you filed.

Reply:
As @stuartmorgan mentioned: if `Teardown` is called on a non-main thread, how are we ensure `surface_pool_->DestroyLayers(jni_facade_);` still happens?

Based on my understanding, we should alway "DestroyLayers", is there a way to destroy them in a non-main thread?
1 idea could be that we asynchronously destroy them on main thread. Then if we creating new layers, they needs to wait for the previous destruction to finish.

Reply:
The embedding can be changed to detect the thread, and run in the right thread, but I don't think it's necessary since the change to rasterizer.cc is preventing that.

Reply:
There would be a chance that `raster_thread_merger_.get() != nullptr` but the current thread is non-main thread right, then `surface_pool_->DestroyLayers(jni_facade_);` won't be executed.

Reply:
what is that case? It's an invalid call to do something like `Teardown(nullptr)`

Reply:
The nullablilty of `raster_thread_merger_.get()` is based on the platforms, so on Android it is always not null. Regardless if threads are merged or not.

One question I just raised now is that when the threads are un-merged, can we say "we are sure there are no layers to be destroyed"? If we can, then your solution is perfect, we just need to add some FML_DCHECK to ensure those for tests.

Reply:
actually, I found a case where my current fix doesn't work. I added a comment explaining the issue, and handled the thread hop directly in Java. PTAL